src/main/kotlin/org/kamiblue/client/module/modules/misc/NoteBot.kt

```diff
@@ -315,4 +320,88 @@ internal object NoteBot : Module(
+
+        channelSettings[0].value = NoteBlockEvent.Instrument.PIANO
+        channelSettings[2].value = NoteBlockEvent.Instrument.BASSDRUM
+        channelSettings[3].value = NoteBlockEvent.Instrument.SNARE
+        channelSettings[4].value = NoteBlockEvent.Instrument.CLICKS
+        channelSettings[5].value = NoteBlockEvent.Instrument.GUITAR
+        channelSettings[6].value = NoteBlockEvent.Instrument.FLUTE
+        channelSettings[7].value = NoteBlockEvent.Instrument.BELL
+        channelSettings[8].value = NoteBlockEvent.Instrument.CHIME
+        channelSettings[9].value = NoteBlockEvent.Instrument.XYLOPHONE
+    }
+
+    fun readNbs(fileName: String): TreeMap<Long, ArrayList<Note>> {
+        val noteSequence = TreeMap<Long, ArrayList<Note>>()
+        val file = File(fileName)
+        val dataInputStream = DataInputStream(FileInputStream(file))
+        var length = dataInputStream.readShort()
+
+        var nbsversion = 0
+        if (length.toInt() == 0) {
+            nbsversion = dataInputStream.readByte().toInt()
+            dataInputStream.readByte().toInt()
+            if (nbsversion >= 3) {
+                length = dataInputStream.readShortCustom()
+            }
+        }
+        dataInputStream.readShortCustom()
+        dataInputStream.skipString()
+        dataInputStream.skipString()
+        dataInputStream.skipString()
+        dataInputStream.skipString()
+        val tempo = dataInputStream.readShortCustom()
+        val timeBetween = 1000 / (tempo / 100).toLong()
+        dataInputStream.skipBytes(23)
+        dataInputStream.skipString()
+        if (nbsversion >= 4) {
+            dataInputStream.skipBytes(4)
+        }
+        var currentTick: Short = -1
+        while (true) {
+            val jump = dataInputStream.readShortCustom()
+            if (jump.toInt() == 0) break
+            currentTick = (currentTick + jump).toShort()
+            var layer: Short = -1
+            while (true) {
+                val jumpLayer: Short = dataInputStream.readShortCustom()
+                if (jumpLayer == 0.toShort()) break
+                layer = (layer + jumpLayer).toShort()
+                val instrument = dataInputStream.readByte()
+                val key = dataInputStream.readByte()
+                if (nbsversion >= 4) {
+                    dataInputStream.readByte() // note block velocity
+                    dataInputStream.readByte() // note block panning
+                    dataInputStream.readShortCustom() // note block pitch
+                }
+                val time = timeBetween * currentTick
+                val note = key % 33
+                noteSequence.getOrPut(time, ::ArrayList).add(Note(note, instrument.coerceIn(0, 15).toInt()))
+            }
+        }
+        return noteSequence
+    }
+
+    private fun DataInputStream.readShortCustom(): Short {
+        val byte1 = readUnsignedByte()
+        val byte2 = readUnsignedByte()
+        return (byte1 + (byte2 shl 8)).toShort()
+    }
```

Comment:
Leave a comment explaining what this is actually doing and what the result means